File: sdk/flutter/packages/mosaic_native_store/ios/Classes/MosaicNativeStorePlugin.swift
@preconcurrency import Flutter
import Foundation
import MosaicSDK
import MosaicStoreKit

public final class MosaicNativeStorePlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = MosaicNativeStorePlugin()
    let channel = FlutterMethodChannel(
      name: "dev.mosaic/native_store/commerce_v1",
      binaryMessenger: registrar.messenger()
    )
    channel.setMethodCallHandler(instance.handle)
    Task { @MainActor in
      await StoreKitProcessBridge.shared.attach(channel)
    }
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
      arguments["codecVersion"] as? Int == 1
    else {
      result(FlutterError(
        code: "codec_version",
        message: "Unsupported Mosaic channel codec.",
        details: nil
      ))
      return
    }
    Task { @MainActor in
      do {
        let bridge = StoreKitProcessBridge.shared
        let value: Any?
        switch call.method {
        case "profile":
          value = try await bridge.profile()
        case "activate":
          value = try await bridge.activate(arguments)
        case "loadProducts":
          value = try await bridge.loadProducts(arguments)
        case "purchase":
          value = try await bridge.purchase(arguments)
        case "restore":
          value = try await bridge.restore()
        case "activeEntitlements":
          value = try await bridge.activeEntitlements()
        case "diagnostics":
          value = try await bridge.diagnostics()
        case "invalidate":
          await bridge.invalidate()
          value = nil
        case "close":
          await bridge.close()
          value = nil
        default:
          throw BridgeError.invalidRequest
        }
        result(value)
      } catch {
        result(FlutterError(
          code: "provider_unavailable",
          message: "The Mosaic StoreKit adapter is unavailable.",
          details: nil
        ))
      }
    }
  }
}

@MainActor
private final class StoreKitProcessBridge {
  static let shared = StoreKitProcessBridge()

  private var provider: MosaicStoreKitProvider?
  private var channels: [FlutterMethodChannel] = []
  private var mappings: [MosaicCommerceProductMapping] = []
  private var updateTask: Task<Void, Never>?

  func attach(_ channel: FlutterMethodChannel) async {
    channels.append(channel)
    do {
      _ = try adapter()
      startUpdateDelivery()
    } catch {
      // The Dart surface reports a stable unavailable outcome on first use.
    }
  }

  func activate(_ arguments: [String: Any]) async throws -> [String: Any] {
    guard arguments["providerId"] as? String == "app_store",
      let reference = arguments["configuration"] as? [String: Any],
      let configurationID = reference["configurationId"] as? String,
      let revision = reference["configurationRevision"] as? String,
      let rawMappings = arguments["mappings"] as? [[String: Any]]
    else { throw BridgeError.invalidRequest }
    let decoded = try rawMappings.map(decodeMapping)
    let provider = try adapter()
    await provider.invalidateLoadedProducts()
    try await provider.install(
      configuration: MosaicCommerceConfigurationReference(
        configurationID: configurationID,
        configurationRevision: revision
      ),
      mappings: decoded
    )
    mappings = decoded
    startUpdateDelivery()
    return ["status": "ready"]
  }

  func profile() throws -> [String: Any] {
    let provider = try adapter()
    return [
      "provider": [
        "id": provider.identity.id,
        "displayName": provider.identity.displayName,
        "adapterVersion": provider.identity.adapterVersion,
      ],
      "capabilities": provider.capabilities.map { capability in
        var value: [String: Any] = [
          "name": capability.name.rawValue,
          "support": capability.support.rawValue,
        ]
        value["reasonCode"] = capability.reasonCode
        return value.compactMapValues { $0 }
      },
      "recoveryMode": "storeSynchronization",
    ]
  }

  func loadProducts(_ arguments: [String: Any]) async throws -> [String: Any] {
    guard let productIDs = arguments["productIds"] as? [String] else {
      throw BridgeError.invalidRequest
    }
    let selected = try productIDs.map { productID in
      guard let mapping = mappings.first(where: { $0.mosaicProductID == productID }) else {
        throw BridgeError.invalidRequest
      }
      return mapping
    }
    let resolved = await adapter().loadProducts(mappings: selected)
    return [
      "products": resolved.map { item in
        guard item.availability.status == .available, let product = item.product,
          let mapping = selected.first(where: { $0.mosaicProductID == item.mosaicProductID })
        else {
          return [
            "mosaicProductId": item.mosaicProductID,
            "availability": item.availability.status.rawValue,
          ]
        }
        return encodeProduct(mapping: mapping, product: product)
      }
    ]
  }

  func purchase(_ arguments: [String: Any]) async throws -> [String: Any] {
    guard let productID = arguments["mosaicProductId"] as? String else {
      throw BridgeError.invalidRequest
    }
    return encodePurchase(await adapter().purchase(mosaicProductID: productID))
  }

  func restore() async throws -> [String: Any] {
    encodeRecovery(await adapter().recover(entitlementMappings: []))
  }

  func activeEntitlements() async throws -> [String: Any] {
    encodeEntitlements(await adapter().activeEntitlements(entitlementMappings: []))
  }

  func diagnostics() async throws -> [String: Any] {
    let value = await adapter().providerDiagnostics()
    return [
      "providerId": value.providerID,
      "health": value.health.rawValue,
      "diagnostics": value.diagnostics.map(encodeDiagnostic),
    ]
  }

  func invalidate() async {
    await provider?.invalidateLoadedProducts()
    mappings = []
  }

  func close() async {
    updateTask?.cancel()
    updateTask = nil
    await provider?.invalidateLoadedProducts()
    provider = nil
    mappings = []
  }

  func accept(
    _ update: MosaicCommerceUpdate
  ) async -> MosaicCommerceUpdateAcceptanceDisposition {
    guard let channel = channels.last else { return .deliveryFailed }
    return await withCheckedContinuation { continuation in
      channel.invokeMethod("commerceUpdate", arguments: encodeUpdate(update)) { response in
        let disposition = (response as? String)
          .flatMap(MosaicCommerceUpdateAcceptanceDisposition.init(rawValue:))
          ?? .deliveryFailed
        continuation.resume(returning: disposition)
      }
    }
  }

  private func adapter() throws -> MosaicStoreKitProvider {
    if let provider { return provider }
    let created = try MosaicStoreKitProvider(
      acceptor: FlutterCommerceUpdateAcceptor()
    )
    provider = created
    return created
  }

  private func startUpdateDelivery() {
    guard updateTask == nil, let provider else { return }
    updateTask = Task { [weak self] in
      let updates = await provider.commerceUpdates
      for await update in updates {
        guard !Task.isCancelled, let self, let channel = self.channels.last else {
          return
        }
        channel.invokeMethod("commerceUpdateReplay", arguments: encodeUpdate(update))
      }
    }
  }
}

private final class FlutterCommerceUpdateAcceptor:
  MosaicCommerceUpdateAcceptor, @unchecked Sendable
{
  func accept(
    _ update: MosaicCommerceUpdate
  ) async throws -> MosaicCommerceUpdateAcceptanceDisposition {
    await StoreKitProcessBridge.shared.accept(update)
  }
}

private func decodeMapping(_ value: [String: Any]) throws
  -> MosaicCommerceProductMapping
{
  guard let productID = value["mosaicProductId"] as? String,
    let mappingID = value["mappingId"] as? String,
    let reference = value["providerProductReference"] as? String,
    let productTypeValue = value["productType"] as? String,
    let productType = MosaicCommerceProductType(rawValue: productTypeValue),
    let grants = value["entitlementKeys"] as? [String],
    let adapter = value["adapterMapping"] as? [String: Any],
    adapter["kind"] as? String == "storeKitProduct"
  else { throw BridgeError.invalidRequest }
  return MosaicCommerceProductMapping(
    mosaicProductID: productID,
    mappingID: mappingID,
    providerProductReference: reference,
    adapterMapping: .storeKitProduct,
    productType: productType,
    entitlementKeys: grants
  )
}

private func encodeProduct(
  mapping: MosaicCommerceProductMapping,
  product: MosaicProduct
) -> [String: Any] {
  var metadata: [String: Any] = [
    "localizedDisplayName": product.title,
    "localizedPrice": product.localizedPrice,
  ]
  metadata["localizedPeriod"] = product.localizedSubscriptionPeriod
  metadata["currencyCode"] = product.currencyCode
  metadata["billingPeriod"] = product.billingPeriod.map(encodePeriod)
  metadata["trial"] = product.trial.map(encodeTrial)
  metadata["introductoryOffer"] = product.introductoryOffer.map(encodeIntroductoryOffer)
  return [
    "mosaicProductId": mapping.mosaicProductID,
    "productType": mapping.productType?.rawValue as Any,
    "entitlementKeys": mapping.entitlementKeys,
    "availability": "available",
    "metadata": metadata.compactMapValues { $0 },
  ]
}

private func encodePeriod(_ value: MosaicCommercePeriod) -> [String: Any] {
  ["unit": value.unit.rawValue, "value": value.value]
}

private func encodeTrial(_ value: MosaicCommerceTrial) -> [String: Any] {
  var result: [String: Any] = ["period": encodePeriod(value.period)]
  result["eligibility"] = value.eligibility?.rawValue
  return result.compactMapValues { $0 }
}

private func encodeIntroductoryOffer(
  _ value: MosaicCommerceIntroductoryOffer
) -> [String: Any] {
  var result: [String: Any] = [
    "localizedPrice": value.localizedPrice,
    "period": encodePeriod(value.period),
    "cycles": value.cycles,
    "paymentMode": value.paymentMode.rawValue,
  ]
  result["eligibility"] = value.eligibility?.rawValue
  return result.compactMapValues { $0 }
}

private func encodePurchase(_ value: MosaicPurchaseResult) -> [String: Any] {
  switch value {
  case .purchased(_, let transaction):
    return ["outcome": "purchased", "transactionReference": transaction as Any]
  case .pending:
    return ["outcome": "pending"]
  case .deferred:
    return ["outcome": "deferred"]
  case .alreadyEntitled:
    return ["outcome": "alreadyEntitled"]
  case .cancelled:
    return ["outcome": "cancelled"]
  case .productUnavailable:
    return ["outcome": "productUnavailable"]
  case .providerUnavailable(_, _, let diagnostic):
    return ["outcome": "providerUnavailable", "diagnostics": [encodeDiagnostic(diagnostic)]]
  case .failed(_, _, let diagnostic):
    return ["outcome": "failed", "diagnostics": [encodeDiagnostic(diagnostic)]]
  }
}

private func encodeRestore(_ value: MosaicRestoreResult) -> [String: Any] {
  switch value {
  case .restored(let entitlements):
    return [
      "outcome": "restored",
      "activeEntitlementKeys": entitlements.map(\.id),
    ]
  case .nothingToRestore: return ["outcome": "nothingToRestore"]
  case .cancelled: return ["outcome": "cancelled"]
  case .providerUnavailable: return ["outcome": "providerUnavailable"]
  case .failed: return ["outcome": "failed"]
  }
}

private func encodeRecovery(
  _ value: MosaicCommerceRecoveryResult
) -> [String: Any] {
  [
    "outcome": value.outcome.rawValue,
    "activeEntitlementKeys": value.activeEntitlements.map(\.id),
    "operationId": value.operationID,
    "providerId": value.providerID,
    "recoveryMode": value.recoveryMode.rawValue,
    "completedAt": utcFormatter.string(from: value.completedAt),
    "diagnostics": value.diagnostics.map(encodeDiagnostic),
  ]
}

private func encodeEntitlements(
  _ value: MosaicActiveEntitlementsResult
) -> [String: Any] {
  switch value {
  case .available(let entitlements):
    return [
      "outcome": "available",
      "activeEntitlementKeys": entitlements.map(\.id),
    ]
  case .unknown: return ["outcome": "unknown"]
  case .providerUnavailable: return ["outcome": "providerUnavailable"]
  case .failed: return ["outcome": "failed"]
  }
}

private func encodeUpdate(_ value: MosaicCommerceUpdate) -> [String: Any] {
  var result: [String: Any] = [
    "updateId": value.id,
    "providerId": value.providerID,
    "mosaicProductId": value.mosaicProductID,
    "configuration": [
      "configurationId": value.configuration.configurationID,
      "configurationRevision": value.configuration.configurationRevision,
    ],
    "outcome": value.outcome.rawValue,
    "activeEntitlementKeys": Array(value.activeEntitlementKeys),
    "occurredAt": utcFormatter.string(from: value.occurredAt),
    "diagnostics": value.diagnostics.map(encodeDiagnostic),
  ]
  result["operationId"] = value.operationID
  result["transactionReference"] = value.transactionReference
  return result.compactMapValues { $0 }
}

private func encodeDiagnostic(_ value: MosaicCommerceDiagnostic) -> [String: Any] {
  var result: [String: Any] = [
    "code": value.code,
    "safeMessage": value.safeMessage,
    "severity": value.severity.rawValue,
    "retryable": value.retryable,
    "correlationId": value.correlationID,
  ]
  result["retryAfterSeconds"] = value.retryAfterSeconds
  result["providerCode"] = value.providerCode
  result["mosaicProductId"] = value.mosaicProductID
  result["recoveryAction"] = value.recoveryAction?.rawValue
  return result.compactMapValues { $0 }
}

private let utcFormatter: ISO8601DateFormatter = {
  let formatter = ISO8601DateFormatter()
  formatter.formatOptions = [.withInternetDateTime]
  return formatter
}()

private enum BridgeError: Error {
  case invalidRequest
}

File: sdk/flutter/packages/mosaic_native_store/android/src/main/kotlin/dev/mosaic/flutter/native_store/MosaicNativeStorePlugin.kt
package dev.mosaic.flutter.native_store

import android.app.Application
import dev.mosaic.sdk.MosaicActiveEntitlementsResult
import dev.mosaic.sdk.MosaicCommerceAdapterMapping
import dev.mosaic.sdk.MosaicCommerceAdapterConfiguration
import dev.mosaic.sdk.MosaicCommerceConfigurationReference
import dev.mosaic.sdk.MosaicCommerceProductMapping
import dev.mosaic.sdk.MosaicCommerceSafeDiagnostic
import dev.mosaic.sdk.MosaicCommerceUpdate
import dev.mosaic.sdk.MosaicCommerceUpdateAcceptance
import dev.mosaic.sdk.MosaicCommerceUpdateAcceptanceDisposition
import dev.mosaic.sdk.MosaicProduct
import dev.mosaic.sdk.MosaicProductLoadResult
import dev.mosaic.sdk.MosaicPurchaseResult
import dev.mosaic.sdk.MosaicRestoreResult
import dev.mosaic.sdk.googleplay.MosaicGooglePlayAdapter
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume

class MosaicNativeStorePlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var application: Application
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var mappings = emptyList<MosaicCommerceProductMapping>()
    private var updateJob: Job? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        application = binding.applicationContext as Application
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
        bridge.attach(application, channel)
        updateJob = scope.launch {
            bridge.adapter(application).commerceUpdates.collect { update ->
                channel.invokeMethod("commerceUpdateReplay", encodeUpdate(update))
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        updateJob?.cancel()
        channel.setMethodCallHandler(null)
        bridge.detach(channel)
        scope.cancel()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val arguments = call.arguments as? Map<*, *>
        if ((arguments?.get("codecVersion") as? Number)?.toInt() != CODEC_VERSION) {
            result.error("codec_version", "Unsupported Mosaic channel codec.", null)
            return
        }
        scope.launch {
            try {
                val value = when (call.method) {
                    "profile" -> bridge.adapter(application).let { adapter ->
                        mapOf(
                            "provider" to mapOf(
                                "id" to adapter.identity.id,
                                "displayName" to adapter.identity.displayName,
                                "adapterVersion" to adapter.identity.adapterVersion,
                            ),
                            "capabilities" to adapter.capabilities.map { capability ->
                                mapOf(
                                    "name" to capability.name,
                                    "support" to capability.support,
                                    "reasonCode" to capability.reasonCode,
                                ).filterValues { it != null }
                            },
                            "recoveryMode" to adapter.recoveryMode,
                        )
                    }
                    "activate" -> activate(arguments)
                    "loadProducts" -> load(arguments)
                    "purchase" -> purchase(arguments)
                    "restore" -> encodeRestore(bridge.adapter(application).restore(emptyList()))
                    "activeEntitlements" ->
                        encodeEntitlements(bridge.adapter(application).activeEntitlements(emptyList()))
                    "diagnostics" -> mapOf(
                        "providerId" to "google_play",
                        "diagnostics" to bridge.adapter(application).diagnostics.map(::encodeDiagnostic),
                    )
                    "invalidate" -> {
                        bridge.adapter(application).invalidateProductHandles()
                        null
                    }
                    "close" -> {
                        bridge.close()
                        null
                    }
                    else -> throw IllegalArgumentException("Unknown Mosaic method.")
                }
                result.success(value)
            } catch (_: Throwable) {
                result.error(
                    "provider_unavailable",
                    "The Mosaic Google Play adapter is unavailable.",
                    null,
                )
            }
        }
    }

    private fun activate(arguments: Map<*, *>): Map<String, Any> {
        require(arguments["providerId"] == "google_play")
        val reference = arguments.map("configuration")
        mappings = arguments.list("mappings").map(::decodeMapping)
        bridge.adapter(application).apply {
            installConfiguration(
                MosaicCommerceAdapterConfiguration(
                    reference = MosaicCommerceConfigurationReference(
                    reference.string("configurationId"),
                    reference.string("configurationRevision"),
                ),
                    mappings = mappings,
                ),
            )
        }
        return mapOf("status" to "ready")
    }

    private suspend fun load(arguments: Map<*, *>): Map<String, Any?> {
        val requested = arguments.list("productIds").map { it as String }
        val selected = requested.map { productId ->
            mappings.single { it.mosaicProductId == productId }
        }
        return when (val response = bridge.adapter(application).loadProducts(selected)) {
            is MosaicProductLoadResult.Loaded -> {
                val productsById = response.products.associateBy(MosaicProduct::id)
                mapOf("products" to selected.map { mapping ->
                    productsById[mapping.mosaicProductId]?.let { product ->
                        encodeProduct(mapping, product)
                    } ?: mapOf(
                        "mosaicProductId" to mapping.mosaicProductId,
                        "availability" to "unavailable",
                    )
                })
            }
            is MosaicProductLoadResult.Unavailable -> mapOf(
                "products" to selected.map {
                    mapOf(
                        "mosaicProductId" to it.mosaicProductId,
                        "availability" to "unavailable",
                    )
                },
            )
        }
    }

    private suspend fun purchase(arguments: Map<*, *>): Map<String, Any?> {
        val productId = arguments.string("mosaicProductId")
        val mapping = mappings.single { it.mosaicProductId == productId }
        return encodePurchase(bridge.adapter(application).purchase(mapping, emptyList()))
    }

    private fun decodeMapping(value: Any?): MosaicCommerceProductMapping {
        val item = value as Map<*, *>
        val adapter = item.map("adapterMapping")
        require(adapter["kind"] == "googlePlayProduct")
        return MosaicCommerceProductMapping(
            mosaicProductId = item.string("mosaicProductId"),
            mappingId = item.string("mappingId"),
            providerProductReference = item.string("providerProductReference"),
            adapterMapping = MosaicCommerceAdapterMapping.GooglePlayProduct(
                adapter["basePlanId"] as? String,
                adapter["offerId"] as? String,
            ),
            productType = item.string("productType"),
            entitlementKeys = item.list("entitlementKeys").map { it as String }.toSet(),
        )
    }

    private companion object {
        const val CHANNEL = "dev.mosaic/native_store/commerce_v1"
        const val CODEC_VERSION = 1
        val bridge = ProcessBridge()
    }
}

private class ProcessBridge {
    private var adapter: MosaicGooglePlayAdapter? = null
    private val channels = linkedSetOf<MethodChannel>()

    @Synchronized
    fun attach(application: Application, channel: MethodChannel) {
        channels += channel
        adapter(application)
    }

    @Synchronized
    fun detach(channel: MethodChannel) {
        channels -= channel
    }

    @Synchronized
    fun adapter(application: Application): MosaicGooglePlayAdapter =
        adapter ?: MosaicGooglePlayAdapter.create(
            application,
            MosaicCommerceUpdateAcceptance { update -> accept(update) },
        ).also { adapter = it }

    suspend fun accept(
        update: MosaicCommerceUpdate,
    ): MosaicCommerceUpdateAcceptanceDisposition {
        val target = synchronized(this) { channels.lastOrNull() }
            ?: return MosaicCommerceUpdateAcceptanceDisposition.DELIVERY_FAILED
        return suspendCancellableCoroutine { continuation ->
            target.invokeMethod("commerceUpdate", encodeUpdate(update), object : MethodChannel.Result {
                override fun success(result: Any?) {
                    val disposition = when (result as? String) {
                        "accepted" -> MosaicCommerceUpdateAcceptanceDisposition.ACCEPTED
                        "alreadyAccepted" ->
                            MosaicCommerceUpdateAcceptanceDisposition.ALREADY_ACCEPTED
                        "rejectedStaleConfiguration" ->
                            MosaicCommerceUpdateAcceptanceDisposition.REJECTED_STALE_CONFIGURATION
                        else -> MosaicCommerceUpdateAcceptanceDisposition.DELIVERY_FAILED
                    }
                    if (continuation.isActive) continuation.resume(disposition)
                }
                override fun error(code: String, message: String?, details: Any?) {
                    if (continuation.isActive) continuation.resume(
                        MosaicCommerceUpdateAcceptanceDisposition.DELIVERY_FAILED,
                    )
                }
                override fun notImplemented() {
                    if (continuation.isActive) continuation.resume(
                        MosaicCommerceUpdateAcceptanceDisposition.DELIVERY_FAILED,
                    )
                }
            })
        }
    }

    @Synchronized
    fun close() {
        adapter?.close()
        adapter = null
    }
}

private fun encodeProduct(
    mapping: MosaicCommerceProductMapping,
    product: MosaicProduct,
): Map<String, Any?> = mapOf(
    "mosaicProductId" to mapping.mosaicProductId,
    "productType" to mapping.productType,
    "entitlementKeys" to mapping.entitlementKeys.toList(),
    "availability" to "available",
    "metadata" to mapOf(
        "localizedDisplayName" to product.title,
        "localizedPrice" to product.localizedPrice,
        "localizedPeriod" to product.subscriptionPeriod,
        "currencyCode" to product.currencyCode,
        "trial" to product.trial?.let(::encodeOffer),
        "introductoryOffer" to product.introductoryOffer?.let(::encodeOffer),
    ).filterValues { it != null },
)

private fun encodePeriod(value: dev.mosaic.sdk.MosaicCommercePeriod) =
    mapOf("unit" to value.unit, "value" to value.value)

private fun encodeOffer(value: dev.mosaic.sdk.MosaicCommerceOffer) = mapOf(
    "localizedPrice" to value.localizedPrice,
    "period" to encodePeriod(value.period),
    "cycles" to value.cycles,
    "paymentMode" to value.paymentMode,
    "eligibility" to value.eligibility,
).filterValues { it != null }

private fun encodePurchase(value: MosaicPurchaseResult): Map<String, Any?> = when (value) {
    is MosaicPurchaseResult.Purchased -> mapOf(
        "outcome" to "purchased",
        "transactionReference" to value.transactionId,
        "activeEntitlementKeys" to value.entitlements.map { it.id },
    )
    is MosaicPurchaseResult.AlreadyEntitled -> mapOf(
        "outcome" to "alreadyEntitled",
        "activeEntitlementKeys" to value.entitlements.map { it.id },
    )
    is MosaicPurchaseResult.Pending -> mapOf("outcome" to "pending")
    is MosaicPurchaseResult.Deferred -> mapOf("outcome" to "deferred")
    is MosaicPurchaseResult.Cancelled -> mapOf("outcome" to "cancelled")
    is MosaicPurchaseResult.ProductUnavailable -> mapOf("outcome" to "productUnavailable")
    is MosaicPurchaseResult.ProviderUnavailable -> mapOf(
        "outcome" to "providerUnavailable",
        "diagnostics" to listOfNotNull(value.diagnostic?.let(::encodeDiagnostic)),
    )
    is MosaicPurchaseResult.Failed -> mapOf(
        "outcome" to "failed",
        "diagnostics" to listOfNotNull(value.diagnostic?.let(::encodeDiagnostic)),
    )
}

private fun encodeRestore(value: MosaicRestoreResult): Map<String, Any?> = when (value) {
    is MosaicRestoreResult.Restored -> mapOf(
        "outcome" to "restored",
        "activeEntitlementKeys" to value.entitlements.map { it.id },
    )
    MosaicRestoreResult.NothingToRestore -> mapOf("outcome" to "nothingToRestore")
    MosaicRestoreResult.Cancelled -> mapOf("outcome" to "cancelled")
    is MosaicRestoreResult.ProviderUnavailable -> mapOf("outcome" to "providerUnavailable")
    is MosaicRestoreResult.Failed -> mapOf("outcome" to "failed")
    is MosaicRestoreResult.Detailed -> mapOf(
        "outcome" to when (value.outcome) {
            dev.mosaic.sdk.MosaicCommerceRecoveryOutcome.RESTORED -> "restored"
            dev.mosaic.sdk.MosaicCommerceRecoveryOutcome.NOTHING_TO_RESTORE ->
                "nothingToRestore"
            dev.mosaic.sdk.MosaicCommerceRecoveryOutcome.CANCELLED -> "cancelled"
            dev.mosaic.sdk.MosaicCommerceRecoveryOutcome.PROVIDER_UNAVAILABLE ->
                "providerUnavailable"
            dev.mosaic.sdk.MosaicCommerceRecoveryOutcome.FAILED -> "failed"
        },
        "activeEntitlementKeys" to value.entitlements.map { it.id },
        "operationId" to value.metadata.operationId,
        "providerId" to value.metadata.providerId,
        "recoveryMode" to value.metadata.recoveryMode,
        "completedAt" to value.metadata.completedAt,
        "diagnostics" to value.metadata.diagnostics.map(::encodeDiagnostic),
    )
}

private fun encodeEntitlements(value: MosaicActiveEntitlementsResult): Map<String, Any?> =
    when (value) {
        is MosaicActiveEntitlementsResult.Available -> mapOf(
            "outcome" to "available",
            "activeEntitlementKeys" to value.entitlements.map { it.id },
        )
        is MosaicActiveEntitlementsResult.Unknown -> mapOf("outcome" to "unknown")
        is MosaicActiveEntitlementsResult.ProviderUnavailable ->
            mapOf("outcome" to "providerUnavailable")
        is MosaicActiveEntitlementsResult.Failed -> mapOf("outcome" to "failed")
    }

private fun encodeUpdate(value: MosaicCommerceUpdate): Map<String, Any?> = mapOf(
    "updateId" to value.updateId,
    "operationId" to value.operationId,
    "providerId" to value.providerId,
    "mosaicProductId" to value.mosaicProductId,
    "configuration" to mapOf(
        "configurationId" to value.configuration.configurationId,
        "configurationRevision" to value.configuration.configurationRevision,
    ),
    "outcome" to when (value.outcome) {
        dev.mosaic.sdk.MosaicCommerceUpdateOutcome.PURCHASED -> "purchased"
        dev.mosaic.sdk.MosaicCommerceUpdateOutcome.PENDING -> "pending"
        dev.mosaic.sdk.MosaicCommerceUpdateOutcome.CANCELLED -> "cancelled"
        dev.mosaic.sdk.MosaicCommerceUpdateOutcome.PROVIDER_UNAVAILABLE ->
            "providerUnavailable"
        dev.mosaic.sdk.MosaicCommerceUpdateOutcome.FAILED -> "failed"
        dev.mosaic.sdk.MosaicCommerceUpdateOutcome.ENTITLEMENTS_CHANGED ->
            "entitlementsChanged"
    },
    "transactionReference" to value.transactionReference,
    "activeEntitlementKeys" to value.activeEntitlements.map { it.id },
    "occurredAt" to value.occurredAt,
    "diagnostics" to value.diagnostics.map(::encodeDiagnostic),
).filterValues { it != null }

private fun encodeDiagnostic(value: MosaicCommerceSafeDiagnostic): Map<String, Any?> = mapOf(
    "code" to value.code,
    "safeMessage" to value.safeMessage,
    "severity" to value.severity,
    "retryable" to value.retryable,
    "retryAfterSeconds" to value.retryAfterSeconds,
    "correlationId" to value.correlationId,
    "providerCode" to value.providerCode,
    "mosaicProductId" to value.mosaicProductId,
    "recoveryAction" to value.recoveryAction,
).filterValues { it != null }

private fun Map<*, *>.string(name: String): String = get(name) as String
private fun Map<*, *>.map(name: String): Map<*, *> = get(name) as Map<*, *>
private fun Map<*, *>.list(name: String): List<*> = get(name) as List<*>
